android/jni/app-android.cpp

```diff
@@ -29,7 +36,351 @@
+			handled = ProcessJoystickMotion(input_event);
+		} else if (source & AINPUT_SOURCE_CLASS_POINTER) {
+			handled = ProcessPointerMotion(input_event);
+		} else if (source & AINPUT_SOURCE_TOUCHPAD) {
+			handled = ProcessTouchpadMotion(input_event);
+		}
+	} else if (AInputEvent_getType(input_event) == AINPUT_EVENT_TYPE_KEY) {
+		if (source & (AINPUT_SOURCE_CLASS_JOYSTICK)) {
+			handled = ProcessButton(input_event, DEVICE_ID_PAD_0);
+		} else {
+			handled = ProcessKey(input_event);
+		}
+	}
+	return handled ? 1 : 0;
+}
+
+// Android native glue: Process the next app command.
+static void engine_handle_cmd(struct android_app* app, int32_t cmd) {
+	struct engine* engine = (struct engine*)app->userData;
+	switch (cmd) {
+	case APP_CMD_SAVE_STATE:
+		ILOG("APP_CMD_SAVE_STATE");
+		// The system has asked us to save our current state. Do so.
+		engine->app->savedState = malloc(sizeof(struct saved_state));
+		*((struct saved_state*)engine->app->savedState) = engine->state;
+		engine->app->savedStateSize = sizeof(struct saved_state);
+		break;
+
+	case APP_CMD_INIT_WINDOW:
+		ILOG("APP_CMD_INIT_WINDOW");
+		// The window is being shown, get it ready.
+		if (engine->app->window) {
+			engine_init_display(engine);
+			NativeInitGraphics();
+			engine_draw_frame(engine);
+		} else {
+			ELOG("APP_CMD_INIT_WINDOW but window == null")
+		}
+		break;
+
+	case APP_CMD_TERM_WINDOW:
+		ILOG("APP_CMD_TERM_WINDOW");
+		// The window is being hidden or closed, clean it up.
+		NativeShutdownGraphics();
+		engine_term_display(engine);
+		break;
+
+	case APP_CMD_GAINED_FOCUS:
+		ILOG("APP_CMD_GAINED_FOCUS");
+		// When our app gains focus, we start monitoring the accelerometer.
+		if (engine->accelerometerSensor) {
+			ASensorEventQueue_enableSensor(engine->sensorEventQueue, engine->accelerometerSensor);
+			// We'd like to get 60 events per second (in us).
+			ASensorEventQueue_setEventRate(engine->sensorEventQueue, engine->accelerometerSensor, (1000L / 60) * 1000);
+		}
+		// We also start animating again.
+		engine->animating = 1;
+		break;
+
+	case APP_CMD_LOST_FOCUS:
+		ILOG("APP_CMD_LOST_FOCUS");
+		// When our app loses focus, we stop monitoring the accelerometer.
+		// This is to avoid consuming battery while not being used.
+		if (engine->accelerometerSensor) {
+			ASensorEventQueue_disableSensor(engine->sensorEventQueue, engine->accelerometerSensor);
+		}
+		// Also stop animating.
+		engine->animating = 0;
+		engine_draw_frame(engine);
+		break;
+	
+	case APP_CMD_INPUT_CHANGED:
+		ILOG("APP_CMD_INPUT_CHANGED");
+		break;
+
+	case APP_CMD_WINDOW_RESIZED:
+		// TODO: Reinitialize EGL?
+		ILOG("APP_CMD_WINDOW_RESIZED");
+		break;
+
+	case APP_CMD_WINDOW_REDRAW_NEEDED:
+		ILOG("APP_CMD_WINDOW_REDRAW_NEEDED");
+		break;
+
+	case APP_CMD_CONTENT_RECT_CHANGED:
+		ILOG("APP_CMD_CONTENT_RECT_CHANGED");
+		break;
+
+	case APP_CMD_CONFIG_CHANGED:
+		ILOG("APP_CMD_CONFIG_CHANGED");
+		break;
+
+	case APP_CMD_LOW_MEMORY:
+		ILOG("APP_CMD_LOW_MEMORY");
```

Comment:
It might be nice to send this over through native, and try to support it across platforms even.

We could do a few things:
 * Display an OSM warning to the user.
 * Cut down on texture cache size (probably largest offender.)
 * Force-disable rewind feature.
 * Disable / ratchet down xBRZ level.

-[Unknown]